### DISCUSSION TITLE
Need help: how read, edit, and write frontmatter?

### DISCUSSION THREAD

**Discussion by jelmerdemaat** [remarkjs/remark#1210] [Q&A]
Hi there! I struggle somewhat to find the correct way to use remark to do the following:

1. Read a markdown file including its frontmatter
2. Edit its frontmatter
3. Write the markdown file to its original location with updated frontmatter

It seems I am at the right place because it looks like one could achieve this with remark, but I'm kind of lost. More precisely, I don't know how to go from the mdast with the updated frontmatter to a written file.

This is where I'm at:

```js
import { read, write } from 'to-vfile';
import { matter } from 'vfile-matter';
import { unified } from 'unified';
import remarkParse from 'remark-parse';
import remarkStringify from 'remark-stringify';
import remarkFrontmatter from 'remark-frontmatter';

function handleYamlFrontmatter() {
    return function (_, file) {
        matter(file);
    };
}

const VFile = await read('testfile.md');

const file = await unified()
      .use(remarkParse)
      .use(remarkStringify)
      .use(remarkFrontmatter)
      .use(handleYamlFrontmatter)
      .process(VFile);

if (file?.data?.matter) {
      file.data.matter.newOrUpdatedPropertyHere = 'testing this 123';

      // So far this works!
      // So... how would I continue here?
}
```

**Comment by ChristianMurphy** [comment]
Welcome @jelmerdemaat! 👋 
A good starting point would be reading https://unifiedjs.com/learn/ to become familiar with syntax trees and how to work with them.
Then look at the nodes frontmatter creates https://github.com/remarkjs/remark-frontmatter, and with a `remark` plugin (https://unifiedjs.com/learn/guide/create-a-plugin/) visit (https://unifiedjs.com/learn/recipe/tree-traversal/) and edit those nodes however you need to.

**Reply by jelmerdemaat** [reply]
Alright! Thanks so much for the super fast response. 

Just to be complete: after that, how would I then go from the edited nodes in the tree to a resulting file: testfile.md ?

**Reply by ChristianMurphy** [reply]
The output VFile, in your example named `file` (`const file = await unified()`), contains the edited Markdown as string/text.
That output virtual file can be written wherever you want to store it: a real file, database, content management system, etc.
Which you can do yourself, or leveraging https://github.com/vfile/to-vfile

**Reply by jelmerdemaat** [reply]
Thanks again! So I tried again using all the documentation links. I still have a few questions:

- I tried [remark-frontmatter](https://github.com/remarkjs/remark-frontmatter) and using [this example](https://github.com/remarkjs/remark-frontmatter#example-frontmatter-as-metadata) it indeed gives me a readable object of the frontmatter of the markdown file as `file.data.matter`. ✔️ 
- I tried to create a plugin and use tree-traversal to edit frontmatter data. But it seems I'm still missing something because:
  -  When I use tree traversal with [visit](https://unifiedjs.com/explore/package/unist-util-visit/#use), I can only traverse the **content** tree, not the **frontmatter** tree. Is this correct?
  - Even so, when I manually edit the given frontmatter data like `file.data.matter.newOrUpdatedPropertyHere = 'testing this 123'` this change will not be reflected back to the `file` (`const file = await unified()`) output. The result of [`file.toString();`](https://github.com/vfile/vfile#vfiletostringencoding) will still be unchanged. Is this correct?

The above would mean I can't change frontmatter data using this method, or that I am still missing something :)

**Reply by remcohaszing** [reply]
`vfile-matter` provides a convenient way to attach frontmatter to the vfile, but when the file is stringified, only the AST is used.

So what you need to do, is modify the AST.

```js
import { read, write } from 'to-vfile';
import { unified } from 'unified';
import remarkParse from 'remark-parse';
import remarkStringify from 'remark-stringify';
import remarkFrontmatter from 'remark-frontmatter';
import { parseDocument } from 'yaml'

function handleYamlFrontmatter() {
    return function (ast) {
        // A simple loop suffices, as frontmatter nodes always exist at the top level.
        for (const node of ast.children) {
            if (node.type !== 'yaml') {
                continue
            }

            // Use parseDocument to preserve YAML formatting and comments
            const doc = parseDocument(node.value)
            // Do your yaml transformations
            doc.setIn(['newOrUpdatedPropertyHere'], 'testing this 123')
            // Set the mdast frontmatter value
            node.value = String(doc)
        }
    };
}

const file = await read('testfile.md');

await unified()
      .use(remarkParse)
      .use(remarkStringify)
      .use(remarkFrontmatter)
      .use(handleYamlFrontmatter)
      .process(file);

console.log(String(file))
```

For more information on modifying YAML, see https://eemeli.org/yaml.

**Reply by wooorm** [reply]
If you would have a string of I dunno, YAML, or TOML, or whatever, in a JavaScript file, and you would use Babel to walk the JS tree it would not visit YAML/TOML/etc either. It’s a completely different thing.

1. a markdown parser does not know about frontmatter by default — it doesn’t exist in CommonMark. You can make a markdown parser understand it as a certain blob, by using an extension. In this case: `remark-frontmatter`. Done: the parser and the compiler now understands your blob.
2. You can use tools that understand that blob as YAML. That’s what you‘re doing with `vfile-matter`. But you may understand that if you have a string of something, and then *parse* some data, and then *change* that data, that doesn’t change anything else. Example: `const a = '{"key": "a"}'; const b = JSON.parse(a); b.key = 'c'; console.log(a) /* not changed */`. *This is what you are missing*.
3. It’s not complex, what you want to do, but you’re missing steps. In this chain, you’re missing **d**: a) parse markdown; b) parse frontmatter as data; c) change data; d) serialize data as frontmatter; e) serialize markdown
